### DISCUSSION TITLE
Compiling for Android with -fPIC

### DISCUSSION THREAD

**Discussion by NLLAPPS** [openssl/openssl#26001] [Q&A]
Hi all, I wonder if someone can help me with compiling for Android with -fPIC. Please note that I am simply putting together what I find on the internet and not a pro.

Up until "3.0.15" below script was able to build fine and I was able to compile OpenSSL in to my binary. I wanted to update to 3.4.0 and changed version variables etc. 

Everything compiled fine but when I tried to compile in to my binary, I started to get "relocation R_ARM_REL32 cannot be used against symbol 'OPENSSL_armcap_P" error. 

After some research, I found that I need to use 

```
export CXXFLAGS="-fPIC"
export CPPFLAGS="-DANDROID -fPIC"
```

As you can see below, I have added them to my script but error did not go away. Just to be sure, I have cleared everything in the build folder before attempting. "export" doesn't seem to work.

Can you see what I am missing? 
NDK version is r27c

Here is my script

`		 
              #!/bin/bash -e
		
		set -e

		# Source variables from config.conf file
		. config.conf

		##############################################################################
		############################      FUNCTIONS     ##############################
		##############################################################################

		function initialSetup {
		    CUR_DIR="$(pwd)"
		    OPENSSL_SRC_PATH="$DOWNLOAD_DIR/${OPENSSL_DIR_NAME}"
		    OPENSSL_TMP_FOLDER="$DOWNLOAD_DIR/tmp/openssl"
		}

		function setupPathsAndExports {
		    NDK_PATH="$DOWNLOAD_DIR/${NDK_DIR_NAME}"
		    LIB_PATH="${OPENSSL_BUILD_OUT_PATH}/libs"
		    LOG_PATH="${OPENSSL_BUILD_OUT_PATH}/logs"
		    # NLL -> Fixes linking error on OpenSSL 3.4.0 and above
		    export CXXFLAGS="-fPIC"
		    export CPPFLAGS="-DANDROID -fPIC"
		    # Export ANDROID_NDK_ROOT and ANDROID_NDK_HOME env var
		    export ANDROID_NDK_HOME=$NDK_PATH
		    export ANDROID_NDK_ROOT=$NDK_PATH
		    # Add toolchains bin directory to PATH
		    TOOLCHAIN_PATH="$NDK_PATH/toolchains/llvm/prebuilt/linux-x86_64"
		    # for some reason it does not need to be exported
		    PATH=$TOOLCHAIN_PATH/bin:$PATH
		}

		function clearBuildDirectory {
		    rm -rf "${OPENSSL_BUILD_OUT_PATH}"
		    mkdir -p "${LIB_PATH}"
		    mkdir -p "${LOG_PATH}"
		}

		function clearTmpDirectory {
		    rm -rf "$OPENSSL_TMP_FOLDER"
		    mkdir -p "$OPENSSL_TMP_FOLDER"
		    cp -r ${OPENSSL_SRC_PATH}/* ${OPENSSL_TMP_FOLDER}
		}

		function getArchitecture {
		    OPENSSL_TARGET_ABI=$1
		    if [ "$OPENSSL_TARGET_ABI" == "armeabi-v7a" ]
		    then
			ARCHITECTURE="android-arm"
		    elif [ "$OPENSSL_TARGET_ABI" == "arm64-v8a" ]
		    then
			ARCHITECTURE="android-arm64"
		    elif [ "$OPENSSL_TARGET_ABI" == "x86" ]
		    then
			# Use "no-asm" arg as specified in Merge Request #28 --- Use Only for x86 ARCH
			ARCHITECTURE="android-x86 no-asm"
		    elif [ "$OPENSSL_TARGET_ABI" == "x86_64" ]
		    then
			ARCHITECTURE="android-x86_64"
		    else
			echo "Unsupported target ABI: $OPENSSL_TARGET_ABI"
			exit 1
		    fi
		}

		##############################################################################
		############################        INIT          ############################
		##############################################################################

		# Initial variables setup
		initialSetup
		# Set final paths and exports
		setupPathsAndExports
		# Clear and recreate the build output directory
		clearBuildDirectory
		# Set clang compiler, instead of gcc by default
		# CC=clang

		##############################################################################
		############################        MAIN          ############################
		##############################################################################

		# Build OpenSSL for each ARCH specified in config.conf
		for arch in "${TARGET_ARCHS[@]}"
		do
		    echo "Configuring OpenSSL for target arch $arch ..."

		    # Clear the tmp source directory
		    clearTmpDirectory
		    # Go to source files
		    cd ${OPENSSL_TMP_FOLDER}

		    OPENSSL_OUTPUT_PATH=$LIB_PATH/$arch

		    # Set the target architecture
		    # Can be android-arm, android-arm64, android-x86 etc
		    ARCHITECTURE="android-arm"
		    getArchitecture $arch
		   
		    # Create Makefile
		    ./Configure $ARCHITECTURE -D__ANDROID_API__=${OPENSSL_TARGET_NDK_LEVEL} enable-pic >> "${LOG_PATH}/${arch}.log" 2>&1

		    # Build Openssl
		    echo "Building OpenSSL Library for Android arch $arch"
		    make >> "${LOG_PATH}/${arch}.log" 2>&1
		    mkdir -p $OPENSSL_OUTPUT_PATH
		    OUTPUT_LIB=$OPENSSL_OUTPUT_PATH/lib
		    mkdir -p $OUTPUT_LIB
		    OUTPUT_INCLUDE=${OPENSSL_OUTPUT_PATH}/include
		    mkdir -p $OUTPUT_INCLUDE
		    cp -RL include/openssl $OUTPUT_INCLUDE

		    # Copy libs to final destination folder
		    cp libssl.a $OUTPUT_LIB
		    cp libcrypto.a $OUTPUT_LIB
		    # Uncomment if you want .so libs (pjsip only needs the static libs)
		    # cp libssl.so $OUTPUT_LIB
		    # cp libcrypto.so $OUTPUT_LIB
		    echo "Build completed! Check output libraries in ${OPENSSL_OUTPUT_PATH}"
		done

		# Remove tmp folder
		rm -rf ${OPENSSL_TMP_FOLDER}
		echo "Finished building OpenSSL! Check output folder: ${OPENSSL_BUILD_OUT_PATH}"

		set +e
`

**Comment by mattcaswell** [comment]
@tom-cosgrove-arm any clue here? With "enable-pic" I would expect this to work?

**Reply by NLLAPPS** [reply]
Hi, just to confirm, I have tried 'enable-pic' with same result

`./Configure $ARCHITECTURE -D__ANDROID_API__=${OPENSSL_TARGET_NDK_LEVEL} enable-pic >> "${LOG_PATH}/${arch}.log" 2>&1`

**Comment by NLLAPPS** [comment]
Here are the relevant parts from generated Make file


` 
     PLATFORM=android-arm
     OPTIONS=-D__ANDROID_API__=21 enable-pic no-acvp-tests no-afalgeng no-asan no-brotli no-brotli-dynamic no-buildtest-c++ no-    crypto-mdebug no-crypto-mdebug-backtrace no-demos no-devcryptoeng no-ec_nistp_64_gcc_128 no-egd no-external-tests no-fips no-    fips-post no-fips-securitychecks no-fuzz-afl no-fuzz-libfuzzer no-h3demo no-jitter no-ktls no-md2 no-msan no-pie no-rc5 no-sctp no-ssl3 no-ssl3-method no-tfo no-trace no-ubsan no-unit-test no-uplink no-weak-ssl-ciphers no-winstore no-zlib no-zlib-dynamic no-zstd no-zstd-dynamic
    CONFIGURE_ARGS=("android-arm", "-D__ANDROID_API__=21", "enable-pic")
    SRCDIR=.
    BLDDIR=.
    FIPSKEY=f455665ffb2d8a43ea2854cbae22ca74560813   
    
    VERSION=3.4.0
    VERSION_NUMBER=3.4.0
    MAJOR=3
    MINOR=4
    SHLIB_VERSION_NUMBER=3
    SHLIB_TARGET=linux-shared

    LIBS=apps/libapps.a libcrypto.a libssl.a providers/libcommon.a \
     providers/libdefault.a providers/liblegacy.a test/libtestutil.a
    SHLIBS=libcrypto.so libssl.so
    SHLIB_INFO="libcrypto.so;;" "libssl.so;;"
    MODULES=engines/capi.so engines/dasync.so engines/loader_attic.so \
        engines/ossltest.so engines/padlock.so providers/legacy.so \
        test/p_minimal.so test/p_test.so

    .........................................
    ##### User defined commands and flags ################################

    # We let the C compiler driver to take care of .s files. This is done in
    # order to be excused from maintaining a separate set of architecture
    # dependent assembler flags. E.g. if you throw -mcpu=ultrasparc at SPARC
    # gcc, then the driver will automatically translate it to -xarch=v8plus
    # and pass it down to assembler.  In any case, we do not define AS or
    # ASFLAGS for this reason.

    CROSS_COMPILE=
    CC=$(CROSS_COMPILE)armv7a-linux-androideabi21-clang
    CXX=$(CROSS_COMPILE)g++
    CPPFLAGS=-D__ANDROID_API__=21 -DANDROID -fPIC
    CFLAGS=-Wall -O3
    CXXFLAGS=-fPIC
    LDFLAGS= 
    EX_LIBS= 
    OBJCOPY=objcopy

    MAKEDEPEND=

    PERL=/usr/bin/perl

    AR=$(CROSS_COMPILE)llvm-ar
    ARFLAGS= rs
    RANLIB=$(CROSS_COMPILE):
    RC= $(CROSS_COMPILE)windres
    RCFLAGS= 

    RM= rm -f
    RMDIR= rmdir
    TAR= tar
    TARFLAGS= 

    BASENAME=       openssl
    NAME=           $(BASENAME)-$(VERSION)
    # Relative to $(SRCDIR)
    TARFILE=        ../$(NAME).tar
`
